### PR TITLE
Check for Signing Root Mismatch When Submitting Proposals and Importing Proposals in Slashing Interchange

### PR DESCRIPTION
**Pull request by rauljordan**
Part of #7813, this PR makes sure we check the following when importing proposals via slashing interchange:

Given signing roots are optional in the EIP standard, we behave as follows:

  For a given block:
    If we have a previous block with the same slot in our history:
      If signing root is empty, we consider that proposer public key as slashable
      If signing root is not empty , then we compare signing roots. If they are different,
      then we consider that proposer public key as slashable.

### CODE REVIEW

File: validator/client/validator.go [hunk 1]
```diff
@@ -507,7 +507,7 @@ func (v *validator) RolesAt(ctx context.Context, slot uint64) (map[[48]byte][]Va
 func (v *validator) UpdateProtections(ctx context.Context, slot uint64) error {
 	attestingPubKeys := make([][48]byte, 0, len(v.duties.CurrentEpochDuties))
 	for _, duty := range v.duties.CurrentEpochDuties {
-		if duty == nil || duty.AttesterSlot != slot {
+		if duty == nil || duty.AttesterSlot != slot-1 {
```
Comment (prestonvanloon) [inline validator/client/validator.go:510]: What's up with this?

File: validator/client/propose_protect.go [hunk 2]
```diff
@@ -33,17 +35,19 @@ func (v *validator) preBlockSignValidations(ctx context.Context, pubKey [48]byte
...
 		}
 		return errors.Wrap(err, "failed to get proposal history")
 	}
-	// If a proposal exists in our history for the slot, we assume it is slashable.
-	// TODO(#7848): Add a more sophisticated strategy where if we indeed have the signing root,
-	// only blocks that have a conflicting signing root with a historical proposal are slashable.
-	if exists {
+
+	// If a proposal exists in our history for the slot, we check the following:
+	// If the signing root is nil, then we consider it slashable. If signing root is not nil,
```
Comment (prestonvanloon) [inline validator/client/propose_protect.go:47]: There is no check if the signing root is nil below. Did you mean if it's zero hash? What if it is `nil`?

File: validator/client/propose_protect.go [hunk 3]
```diff
@@ -33,17 +35,19 @@ func (v *validator) preBlockSignValidations(ctx context.Context, pubKey [48]byte
...
+	// we check if it is different than the incoming block's signing root. If that is the case,
+	// we consider that proposal slashable.
+	if exists && (prevSigningRoot == params.BeaconConfig().ZeroHash || prevSigningRoot != signingRoot) {
```
Comment (prestonvanloon) [inline validator/client/propose_protect.go:50]: Shouldn't we stop a proposal if a block proposal was already submitted? 
Why allow another block to be proposed, even if it has exactly the same information? 

Whatever benefit there is, I dont know if it's worth the risk of proposing a block when the validator has already proposed one.

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Can you update the title? There are changes in the proposal logic rather than changes only in the import logic.

File: validator/client/propose_protect.go (hunk 3 above)
Comment (rauljordan) [reply validator/client/propose_protect.go:50]: This is totally allowed. The only slashable offense is if the signing root does not match but the slot matches

File: validator/client/propose_protect.go (hunk 2 above)
Comment (rauljordan) [reply validator/client/propose_protect.go:47]: Cannot be nil because the return type is [32]byte

File: validator/client/propose_protect.go (hunk 3 above)
Comment (rauljordan) [reply validator/client/propose_protect.go:50]: Also this is a test case for the EIP spec tests. We expect a block with same slot and same signing root to go through ok

File: beacon-chain/state/stateutil/attestations.go [hunk 4]
```diff
@@ -84,66 +84,6 @@ func marshalAttestationData(data *ethpb.AttestationData) []byte {
 	return enc
 }
 
-func attestationRoot(hasher htrutils.HashFn, att *ethpb.Attestation) ([32]byte, error) {
```
Comment (rauljordan) [inline beacon-chain/state/stateutil/attestations.go:87]: Deep source was complaining of unused code

File: validator/client/propose.go [hunk 5]
```diff
@@ -101,7 +94,23 @@ func (v *validator) ProposeBlock(ctx context.Context, slot uint64, pubKey [48]by
...
+	}
+
+	if err := v.preBlockSignValidations(ctx, pubKey, b, signingRoot); err != nil {
+		log.WithFields(
+			blockLogFields(pubKey, b, nil),
+		).WithError(err).Error("Failed block slashing protection check")
+		return
+	}
+
+	if err := v.postBlockSignUpdate(ctx, pubKey, blk, signingRoot); err != nil {
```
Comment (prestonvanloon) [inline validator/client/propose.go:113]: Maybe something for another PR, but these should just be merged into 1 method.

File: validator/client/propose_protect.go [hunk 6]
```diff
@@ -6,50 +6,58 @@ import (
...
+	"github.com/prysmaticlabs/prysm/shared/params"
 	"github.com/sirupsen/logrus"
 )
 
 var failedPreBlockSignLocalErr = "attempted to sign a double proposal, block rejected by local protection"
 var failedPreBlockSignExternalErr = "attempted a double proposal, block rejected by remote slashing protection"
 var failedPostBlockSignErr = "made a double proposal, considered slashable by remote slashing protection"
 
-func (v *validator) preBlockSignValidations(ctx context.Context, pubKey [48]byte, block *ethpb.BeaconBlock) error {
+func (v *validator) preBlockSignValidations(
```
Comment (prestonvanloon) [inline validator/client/propose_protect.go:19]: Again, maybe for another PR. Would like some documentation on what this method does here.

**Review by prestonvanloon — APPROVED** [review]

File: validator/client/propose.go (hunk 5 above)
Comment (rauljordan) [reply validator/client/propose.go:113]: Agreed, but out of scope. Will open an issue

**Pull request merged**